MINOR: Replace BrokerMetadataListener with MetadataLoader

cmccabe opened the pull request:
MINOR: Replace BrokerMetadataListener with MetadataLoader

Replace BrokerMetadataListener with MetadataLoader. BrokerMetadataListener was, in a sense, an
early prototype of MetadataLoader.It constructed a MetadataImage object from records supplied by
the raft layer.

MetadataLoader is a rewrite of this code with several improvements.

- MetadataLoadder is in Java, and not part of the core gradle module.

- MetadataLoadder is not broker-specific. This is especially useful in combined mode where it means
  we do not have to have a separate, controller-specific code path for creating MetadataImage
  objects.

- MetadataLoader supports multiple publishers rather than a single publisher.

- Publisher can be dynamically added and removed from MetadataLoader.

cmccabe committed:
MINOR: Replace BrokerMetadataListener with MetadataLoader

Replace BrokerMetadataListener with MetadataLoader. BrokerMetadataListener was, in a sense, an
early prototype of MetadataLoader.It constructed a MetadataImage object from records supplied by
the raft layer.

MetadataLoader is a rewrite of this code with several improvements.

- MetadataLoadder is in Java, and not part of the core gradle module.

- MetadataLoadder is not broker-specific. This is especially useful in combined mode where it means
  we do not have to have a separate, controller-specific code path for creating MetadataImage
  objects.

- MetadataLoader supports multiple publishers rather than a single publisher.

- Publisher can be dynamically added and removed from MetadataLoader.

mumrah commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -362,16 +326,13 @@ class BrokerServer(
       )
       lifecycleManager.start(
-        () => metadataListener.highestMetadataOffset,
+        () => sharedServer.loader.lastAppliedOffset(),
```
Just want to check that the semantics are the same here (since the terminology and code is slightly different). Before, we were sending `highestMetadataOffset` which is updated as we replay messages:

```scala
        _highestOffset = lastCommittedOffset.getOrElse(batch.baseOffset() + index)
        try {
          delta.replay(messageAndVersion.message())
        } catch {
```

We don't actually call the publishers until after we have replayed the whole batch.

With the new metadata loader infrastructure, we update the `lastAppliedOffset` after we have applied the whole batch and called the publishers.

cmccabe replied:
This is a good question. It is indeed a slightly different semantic, and the PR description should call that out.

As you said, previously we were not waiting for the updates we loaded to actually be published. Now we are. The intention was always to wait for the updates to actually be propagated before declaring the broker "caught up" so this is really fixing a bug. I'll make a note in the description.

mumrah reviewed (commented):
@cmccabe, nice to see consolidation of metadata handling on the controller and broker 😁 

Left one question inline

cmccabe committed:
fix ServerShutdownTest

fix ServerShutdownTest: need to ensure that if we can't contact the
quorum we fail

cmccabe committed:
update comment

mumrah reviewed (approved):
Thanks for adding the commentary in BrokerServer, it's helpful to explain all the various start up phases.

LGTM :shipit:
